sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -263,4 +280,107 @@ case class InsertIntoHadoopFsRelationCommand(
       }
     }.toMap
   }
+
+  /**
+   * Detect the conflict when there are several InsertIntoHadoopFsRelation operations
+   * write to same partition in the same table or a non-partitioned table concurrently.
+   */
+  private def detectConflict(
```

Comment:
All this is happening before the `committer.setupJob` call, right?

Doesn't that mean that two applications may concurrently pass this test, and then mess up each other when they actually start running the job?

It feels like there need to be a fancier protocol here to define who can continue past this point, which would actually involve writing to the underlying file system. e.g. by creating the needed staging directory(ies) before doing this check (throwing an error if the directory already exists).

Reply:
yes, you are right. Thanks.

Reply:
fixed